File: 백준/Silver/17212. 달나라 토끼를 위한 구매대금 지불 도우미/달나라 토끼를 위한 구매대금 지불 도우미.cc
#include <iostream>
using namespace std;

int dp[4][100001];
int a[4] = {1,2,5,7};

int main() {
    int n;cin>>n;
    
    for(int i=0;i<4;i++){
        for(int j=1;j<=n;j++){
            dp[i][j] = 987654321;
        }
    }
    
    for(int i=0;i<4;i++){
        for(int j=1;j<=n;j++){
            if(i==0) dp[i][j] = j;
            else {
                if(j < a[i]) dp[i][j] = dp[i-1][j];
                else dp[i][j] = min(dp[i][j-a[i]]+1,dp[i-1][j]);
            }
        }
    }
    
    cout<<dp[3][n];
}

File: 백준/Silver/17212. 달나라 토끼를 위한 구매대금 지불 도우미/달나라 토끼를 위한 구매대금 지불 도우미.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());

		int[] dp = new int[100001];

		dp[1] = 1;
		for (int i = 2; i <= n; i++) {
			dp[i] = dp[i - 1] + 1;
			dp[i] = Math.min(dp[i], dp[i - 2] + 1);
			if (i >= 5) {
				dp[i] = Math.min(dp[i], dp[i - 5] + 1);
			}
			if (i >= 7) {
				dp[i] = Math.min(dp[i], dp[i - 7] + 1);
			}
		}

		System.out.println(dp[n]);
	}
}